Make the firmware remotely updatable

bschwind opened the pull request:
@skywhale please review

Here's the basic summary of the changes:

* Add a `Bootload` command to tell the device to go into bootloader mode
* Write some "sentinel" value in an STM32 backup register (special registers that can survive system resets)
* Request a system reset
* On firmware startup, check the backup register for the sentinel value.
* If the sentinel value is present, reset it to `0` and jump to the bootloader
* If the sentinel value is not present, continue in the firmware as normal

The reason for this little dance instead of just jumping directly into the bootloader is that we would otherwise have to reset all device peripherals back to their default state, including USB devices, timers, etc.

It's much easier to simply reset and jump to the bootloader before we configure any hardware devices.

References:

* STM32F4 Backup Register Usage
https://www.st.com/resource/en/reference_manual/dm00119316-stm32f411xc-e-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf

<img width="1095" alt="Screen Shot 2021-05-21 at 3 11 46 PM" src="https://user-images.githubusercontent.com/458432/119090676-e1944600-ba46-11eb-96d7-3fedeafe5507.png">

* STM32 Bootloader Application Note AN2606
https://www.st.com/resource/en/application_note/cd00167594-stm32-microcontroller-system-memory-boot-mode-stmicroelectronics.pdf

<img width="700" alt="Screen Shot 2021-05-17 at 3 07 15 PM" src="https://user-images.githubusercontent.com/458432/119090731-f1ac2580-ba46-11eb-9224-7bbf6f58d2b3.png">

skywhale commented on src/main.rs:
```diff
@@ -179,6 +187,10 @@ fn main() -> ! {
                 },
+                Command::Bootload => {
+                    led.set_high().unwrap();
+                    bootload::request_bootloader();
```
`return bootload::request_bootloader();` to make it clear that this is the end of the loop?

bschwind replied:
`request_bootloader` returns the [never type](https://doc.rust-lang.org/reference/types/never.html) which means it will never return control flow to this call, which means the `return` code itself will not run. I actually get a warning about unreachable code if I add a `return` here.

skywhale commented on src/main.rs:
```diff
@@ -64,6 +68,7 @@ fn main() -> ! {
 
     // Set up the LED (C13).
     let mut led = gpioc.pc13.into_push_pull_output();
+    led.set_high().unwrap();
```
do you want to create a simple `Led` wrapper? `set_high()` is to turn off, I suppose, but that's confusing at glance

bschwind replied:
I'll do this in a followup PR.

skywhale commented on src/bootload.rs:
```diff
@@ -0,0 +1,70 @@
+    rcc.bdcr.write(|w| w.rtcen().bit(true));
+}
+
+fn disable_backup_domain(dp: &stm32::Peripherals) {
```
could you comment that we are not reverting all the state changes in `enable_backup_domain`, as we are either `sys_reset()` or `boatload()` after this?

bschwind replied:
Let me try doing the exact reverse of `enable_backup_domain` to see if it still works, feels cleaner that way.

bschwind replied:
Seems to work well, I'll reverse all the steps done in `enable_backup_domain`.

skywhale commented on src/bootload.rs:
```diff
@@ -0,0 +1,70 @@
+    cortex_m::asm::dsb();
+
+    // Set the DBP bit in the Section 5.4.1 to enable access to the backup domain
+    pwr.cr.write(|w| w.dbp().bit(true));
```
super nit: does the order of operation matter? if not, making this the last call of `enable_backup_domain()` feels more symmetric?

bschwind replied:
I'm following the datasheet here, and I've seen other implementations do it in this order as well. Here's a C implementation of a similar thing:

```c
int8_t write_to_backup_rtc( uint32_t *data, uint16_t bytes, uint16_t offset ) {
  const uint16_t backup_size = 80;
  volatile uint32_t* base_addr = &(RTC->BKP0R);
  uint16_t i;
  if( bytes + offset >= backup_size ) {
    /* ERROR : the last byte is outside the backup SRAM region */
    return -1;
  } else if( offset % 4 || bytes % 4 ) {
    /* ERROR: data start or num bytes are not word aligned */
    return -2;
  } else {
    bytes >>= 2;      /* divide by 4 because writing words */
  }
  /* disable backup domain write protection */
  RCC_APB1PeriphClockCmd(RCC_APB1Periph_PWR, ENABLE);   // set RCC->APB1ENR.pwren
  PWR_BackupAccessCmd(ENABLE);                          // set PWR->CR.dbp = 1;
  for( i = 0; i < bytes; i++ ) {
    *(base_addr + offset + i) = *(data + i);
  }
  PWR_BackupAccessCmd(DISABLE);                     // reset PWR->CR.dbp = 0;
  // consider also disabling the power peripherial?
  return 0;
}

int8_t read_from_backup_rtc( uint32_t *data, uint16_t bytes, uint16_t offset ) {
  const uint16_t backup_size = 80;
  volatile uint32_t* base_addr = &(RTC->BKP0R);
  uint16_t i;
  if( bytes + offset >= backup_size ) {
    /* ERROR : the last byte is outside the backup SRAM region */
    return -1;
  } else if( offset % 4 || bytes % 4 ) {
    /* ERROR: data start or num bytes are not word aligned */
    return -2;
  } else {
    bytes >>= 2;      /* divide by 4 because writing words */
  }
  /* read should be 32 bit aligned */
  for( i = 0; i < bytes; i++ ) {
    *(data + i) = *(base_addr + offset + i);
  }
  return 0;
}
```

bschwind replied:
> making this the last call of enable_backup_domain() feels more symmetric?

Feels more symmetric with respect to what?

skywhale reviewed (approved):
LGTM

Looking great, thank you for figuring this out!

skywhale reviewed (approved):
LGTM++

bschwind commented:
@skywhale could I get one more look at this? I changed the USB code a bit to initialize the USB device right before the loop so it's less likely to miss any USB communication from the host, and I also added some code to reset the bus before starting up.

Occasionally after uploading firmware via DFU, the tonari serial device would not show up on the host. This forced reset condition seems to fix it. I have jumped to the bootloader, uploaded new firmware, and reset dozens of times without failure on my laptop. I'll test it on linux as well.

skywhale commented on src/main.rs:
```diff
@@ -148,6 +121,43 @@ fn main() -> ! {
+    // after booting up.
+    let mut delay = hal::delay::Delay::new(cp.SYST, clocks);
+    let mut usb_pin_d_plus = gpioa.pa12.into_push_pull_output();
+    usb_pin_d_plus.set_low().unwrap();
```
where are you resetting back? is it done by `UsbDeviceBuilder` later in the code?

bschwind replied:
Yep, once we pass ownership of this pin to the USB structs, they will take care of setting the pins to the proper state for communication.

strohel commented on src/bootload.rs:
```diff
@@ -0,0 +1,80 @@
+    rcc.bdcr.write(|w| w.rtcen().bit(true));
+}
+
+fn disable_backup_domain(dp: &stm32::Peripherals) {
```
Very optional: if we want to get super fancy, we can present {enable,disable}_backup_domain() as a RAII-style guard (implement the disable in its drop()). Alternative: as some closure-taking function `in_backup_domain(dp: &stm32::Peripherals, f: impl FnOnce)`. Not sure if that wouldn't be over-engineering though.

bschwind replied:
Yes I thought about this, but decided it was a bit too much for such a simple use case.

I think where this would be useful is in the `stm32f4xx-hal` crate itself (or maybe a more general `stm32` crate), where they have more Rust-y interfaces for these things.

strohel replied:
:+1:

strohel reviewed (commented):
I cannot claim I understand the details, but high-level wise this was nicely readable.
